drivers/video/video_esp32_dvp.c

```diff
@@ -0,0 +1,463 @@
+
+	if (data->active_vbuf == NULL) {
+		LOG_WRN("Frame dropped. No buffer available");
+		return;
+	}
+
+	video_esp32_reload_dma(data);
+}
+
+void video_esp32_dma_rx_done(const struct device *dev, void *user_data, uint32_t channel,
+			     int status)
+{
+	struct video_esp32_data *data = user_data;
+	int ret = 0;
+
+	if (status == DMA_STATUS_BLOCK) {
+		LOG_DBG("received block");
+		return;
+	}
+
+	if (status != DMA_STATUS_COMPLETE) {
+		LOG_ERR("DMA error: %d", status);
+		return;
+	}
+
+	if (data->active_vbuf == NULL) {
+		LOG_ERR("No video buffer available. Enque some buffers first.");
+		return;
+	}
+
+	k_fifo_put(&data->fifo_out, data->active_vbuf);
+	data->active_vbuf = NULL;
+}
+
+static int video_esp32_reload_dma(struct video_esp32_data *data)
+{
+	const struct video_esp32_config *cfg = data->config;
+	int ret = 0;
+
+	if (data->active_vbuf == NULL) {
+		LOG_ERR("No video buffer available. Enque some buffers first.");
+		return -EAGAIN;
+	}
+
+	ret = dma_reload(cfg->dma_dev, cfg->rx_dma_channel, 0, (uint32_t)data->active_vbuf->buffer,
+			 data->active_vbuf->bytesused);
+	if (ret) {
+		LOG_ERR("Unable to reload DMA (%d)", ret);
+		return ret;
+	}
+
+	ret = dma_start(cfg->dma_dev, cfg->rx_dma_channel);
+	if (ret) {
+		LOG_ERR("Unable to start DMA (%d)", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int video_esp32_stream_start(const struct device *dev)
+{
+	const struct video_esp32_config *cfg = dev->config;
+	struct video_esp32_data *data = dev->data;
+	struct dma_status dma_status = {0};
+	struct dma_config dma_cfg = {0};
+	struct dma_block_config *dma_block_iter = data->dma_blocks;
+	uint32_t buffer_size = 0;
+	unsigned int key = irq_lock();
+	int error = 0;
+
+	LOG_DBG("Start streaming");
+
+	error = dma_get_status(cfg->dma_dev, cfg->rx_dma_channel, &dma_status);
+
+	if (error) {
+		LOG_ERR("Unable to get Rx status (%d)", error);
+		goto unlock;
+	}
+
+	if (dma_status.busy) {
+		LOG_ERR("Rx DMA Channel %d is busy", cfg->rx_dma_channel);
+		error = -EBUSY;
+		goto unlock;
+	}
+
+	data->active_vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
+	if (!data->active_vbuf) {
+		LOG_ERR("No enqueued video buffers available.");
+		error = -EAGAIN;
+		goto unlock;
+	}
+
+	buffer_size = data->active_vbuf->bytesused;
+	memset(data->dma_blocks, 0, sizeof(data->dma_blocks));
+	for (int i = 0; i < CONFIG_DMA_ESP32_MAX_DESCRIPTOR_NUM; ++i) {
+		dma_block_iter->dest_address =
+			(uint32_t)data->active_vbuf->buffer + i * VIDEO_ESP32_DMA_BUFFER_MAX_SIZE;
+		if (buffer_size < VIDEO_ESP32_DMA_BUFFER_MAX_SIZE) {
+			dma_block_iter->block_size = buffer_size;
+			dma_block_iter->next_block = NULL;
+			dma_cfg.block_count = i + 1;
+			break;
+		}
+		dma_block_iter->block_size = VIDEO_ESP32_DMA_BUFFER_MAX_SIZE;
+		dma_block_iter->next_block = dma_block_iter + 1;
+		dma_block_iter++;
+		buffer_size -= VIDEO_ESP32_DMA_BUFFER_MAX_SIZE;
+	}
+
+	if (dma_block_iter->next_block) {
+		LOG_ERR("Not enough descriptors available. Increase DMA_ESP32_DESCRIPTOR_NUM");
```

Comment:
I noticed there is neither irq_unlock, not goto_unlock which may cause irq state conflict.

Reply:
good spot